sidekick_core/lib/src/sidekick_context.dart

```diff
@@ -85,74 +134,183 @@ class SidekickContext {
+        );
+      }
+      return result;
+    });
+  }
 
-        current = current.parent;
+  /// The directory where the [entryPoint] is located
+  ///
+  /// This directory is considered the project root.
+  static Directory get projectRoot {
+    if (env.exists(_envEntryPointHome)) {
+      // CLI is called via entryPoint
+      final injectedEntryPointPath = env[_envEntryPointHome];
+      if (injectedEntryPointPath == null || injectedEntryPointPath.isBlank) {
+        throw 'Injected entryPoint was not set (env.$_envEntryPointHome)';
       }
-      throw "Can't find entryPoint $entryPointName from ${sidekickPackageDir.path}";
+      final Directory dir = Directory(injectedEntryPointPath);
+      if (!dir.existsSync()) {
+        throw 'Injected entryPoint does not exist ${dir.absolute.path}';
+      }
+      return dir;
     }
+    return entryPoint.file.parent;
   }
 
   /// The git repository root the [sidekickPackage] is located in
-  static Directory get repository {
-    bool isGitDir(Directory dir) => dir.directory('.git').existsSync();
+  @Deprecated('Use projectRoot instead')
+  static Repository get repository {
+    // TODO remove repository? It's not actually required when we define
+    //  the entryPoint to be the root of the project and start package discovery
+    //  from there.
+    return _cache.getOrCreate('findRepository', _findRepository);
+  }
 
-    final gitRoot = sidekickPackageDir.findParent(isGitDir);
+  static Repository _findRepository() {
+    bool isGitDir(Directory dir) => dir.directory('.git').existsSync();
 
+    final projectRoot = entryPoint.file.parent;
+    final gitRoot = projectRoot.findParent(isGitDir);
     if (gitRoot == null) {
-      throw 'Could not find the root of the repository. Searched in '
-          '${sidekickPackageDir.absolute.path}';
+      throw 'Could not find the root of the repository from ${projectRoot.path}';
     }
+    return Repository(root: gitRoot);
+  }
+}
 
-    return gitRoot;
+SidekickContextCache get internalSidekickContextCache {
+  return SidekickContext._cache;
+}
+
+set internalSidekickContextCache(SidekickContextCache value) {
+  SidekickContext._cache = value;
+}
+
+abstract class SidekickContextCache {
+  T getOrCreate<T extends Object>(Object key, T Function() create);
+  factory SidekickContextCache() = _InMemoryCache;
+  factory SidekickContextCache.noCache() = _NoCache;
+}
+
+class _NoCache implements SidekickContextCache {
+  @override
+  T getOrCreate<T extends Object>(Object key, T Function() create) {
+    return create();
+  }
+}
+
+class _InMemoryCache implements SidekickContextCache {
+  final Map<Object, Object> _map = {};
+  @override
+  T getOrCreate<T extends Object>(Object key, T Function() create) {
+    final value = _map[key] as T?;
+    if (value != null) {
+      return value;
+    }
+    final newValue = create();
+    _map[key] = newValue;
+    return newValue;
   }
 }
+
+class ProjectDiscoveryResult {
```

Comment:
Should we make `ProjectDiscoveryResult` private? It's only being used in this file